File: python/deobfuscator.py
#!/usr/bin/env python3
import sys
import re
import base64
import hashlib


def extract_payload(path):
    text = open(path, 'r', encoding='utf-8').read()
    # base64 payload (supports triple-quoted or single/double-quoted)
    m = re.search(r'_b64\s*=\s*"""(.+?)"""', text, re.S)
    if not m:
        m = re.search(r"_b64\s*=\s*([\'\"])" + r"(.+?)" + r"\1", text, re.S)
    if not m:
        raise ValueError('No embedded base64 payload found')
    # For the first pattern group 1 is the content; for the second, group 2 is the content
    if m.lastindex == 1:
        b64 = m.group(1)
    else:
        b64 = m.group(2)

    m2 = re.search(r'_salt_hex\s*=\s*"([0-9a-fA-F]+)"', text)
    if not m2:
        m2 = re.search(r"_salt_hex\s*=\s*([\'\"])" + r"([0-9a-fA-F]+)" + r"\1", text)
    salt = bytes.fromhex(m2.group(1) if m2 and m2.lastindex==1 else (m2.group(2) if m2 else '')) if m2 else None

    m3 = re.search(r"_iters\s*=\s*(\d+)", text)
    iters = int(m3.group(1)) if m3 else None

    return b64, salt, iters


def xor_bytes(data: bytes, key_bytes: bytes) -> bytes:
    return bytes([b ^ key_bytes[i % len(key_bytes)] for i, b in enumerate(data)])


def main():
    if len(sys.argv) < 3:
        print('Usage: deobfuscator.py obfuscated.py password [out.py]')
        sys.exit(2)
    obf = sys.argv[1]
    password = sys.argv[2]
    out = sys.argv[3] if len(sys.argv) > 3 else obf + '.decoded.py'
    b64, salt, iters = extract_payload(obf)
    raw = base64.b64decode(b64)

    if salt is None or iters is None:
        print('Missing salt/iterations in obfuscated file; cannot derive key')
        sys.exit(3)

    key_bytes = hashlib.pbkdf2_hmac('sha256', password.encode(), salt, iters, dklen=32)
    orig = xor_bytes(raw, key_bytes)
    with open(out, 'wb') as f:
        f.write(orig)
    print('Wrote decoded file to', out)


if __name__ == '__main__':
    main()

File: go/deobfuscator.go
package main

import (
    "encoding/base64"
    "encoding/hex"
    "flag"
    "fmt"
    "io/ioutil"
    "os"
    "regexp"
    "strconv"
    "crypto/hmac"
    "crypto/sha256"
)

func pbkdf2(password, salt []byte, iter, keyLen int) []byte {
    hLen := 32 // sha256
    blocks := (keyLen + hLen - 1) / hLen
    dk := make([]byte, 0, blocks*hLen)
    for b := 1; b <= blocks; b++ {
        mac := hmac.New(sha256.New, password)
        mac.Write(salt)
        mac.Write([]byte{byte(b >> 24), byte(b >> 16), byte(b >> 8), byte(b)})
        u := mac.Sum(nil)
        t := make([]byte, len(u))
        copy(t, u)
        for i := 1; i < iter; i++ {
            mac = hmac.New(sha256.New, password)
            mac.Write(u)
            u = mac.Sum(nil)
            for j := 0; j < len(t); j++ {
                t[j] ^= u[j]
            }
        }
        dk = append(dk, t...)
    }
    return dk[:keyLen]
}

func xor(data, key []byte) []byte {
    out := make([]byte, len(data))
    for i := range data {
        out[i] = data[i] ^ key[i%len(key)]
    }
    return out
}

func main() {
    in := flag.String("in", "", "obfuscated go wrapper file")
    out := flag.String("out", "decoded.go", "output file")
    pass := flag.String("pass", "", "password")
    flag.Parse()
    if *in == "" || *pass == "" {
        fmt.Println("Usage: deobfuscator -in wrapper.go -pass password [-out decoded.go]")
        os.Exit(2)
    }
    text, err := ioutil.ReadFile(*in)
    if err != nil {
        fmt.Println("read error:", err)
        os.Exit(1)
    }
    // find base64 payload
    reB64 := regexp.MustCompile("b64 := `([A-Za-z0-9+/=\\n]+)`")
    m := reB64.FindSubmatch(text)
    if m == nil {
        fmt.Println("no payload found")
        os.Exit(3)
    }
    b64 := string(m[1])
    // find salt hex
    reSalt := regexp.MustCompile("saltHex := \"([0-9a-fA-F]+)\"")
    ms := reSalt.FindSubmatch(text)
    if ms == nil {
        fmt.Println("no salt found")
        os.Exit(4)
    }
    salt, _ := hex.DecodeString(string(ms[1]))
    // find iters
    reIters := regexp.MustCompile("iters := (\\d+)")
    mi := reIters.FindSubmatch(text)
    if mi == nil {
        fmt.Println("no iterations found")
        os.Exit(5)
    }
    iters, _ := strconv.Atoi(string(mi[1]))

    raw, _ := base64.StdEncoding.DecodeString(b64)
    key := pbkdf2([]byte(*pass), salt, iters, 32)
    dec := xor(raw, key)
    if err := ioutil.WriteFile(*out, dec, 0644); err != nil {
        fmt.Println("write error:", err)
        os.Exit(6)
    }
    fmt.Println("Wrote decoded file:", *out)
}
